iceoryx_posh/test/moduletests/test_popo_server_port_user.cpp

```diff
@@ -0,0 +1,732 @@
+    });
+
+    EXPECT_TRUE(detectedError.has_value());
+}
+
+TEST_F(ServerPort_test, FreeResponseWithWithNullptrResponseHeaderCallsTheErrorHandler)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ecb40c4d-7b95-4780-9b51-ac1708830453");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    iox::cxx::optional<iox::Error> detectedError;
+    auto errorHandlerGuard = iox::ErrorHandler::setTemporaryErrorHandler(
+        [&](const iox::Error error, const std::function<void()>, const iox::ErrorLevel errorLevel) {
+            EXPECT_THAT(error, Eq(iox::Error::kEXPECTS_ENSURES_FAILED));
+            EXPECT_THAT(errorLevel, Eq(iox::ErrorLevel::FATAL));
+            detectedError.emplace(error);
+        });
+
+    sut.portUser.freeResponse(nullptr);
+
+    EXPECT_TRUE(detectedError.has_value());
+}
+
+// END freeResponse tests
+
+// BEGIN sendResponse tests
+
+TEST_F(ServerPort_test, SendResponseWithWithNullptrResponseHeaderCallsTheErrorHandler)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8d0b0a7f-d1ab-4249-b885-cbc6429eab83");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    iox::cxx::optional<iox::Error> detectedError;
+    auto errorHandlerGuard = iox::ErrorHandler::setTemporaryErrorHandler(
+        [&](const iox::Error error, const std::function<void()>, const iox::ErrorLevel errorLevel) {
+            EXPECT_THAT(error, Eq(iox::Error::kEXPECTS_ENSURES_FAILED));
+            EXPECT_THAT(errorLevel, Eq(iox::ErrorLevel::FATAL));
+            detectedError.emplace(error);
+        });
+
+    sut.portUser.sendResponse(nullptr);
+
+    EXPECT_TRUE(detectedError.has_value());
+}
+
+TEST_F(ServerPort_test, SendResponseWithWithoutOfferReleasesTheChunkToTheMempool)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "dc4e31b1-18bf-4c42-9084-dd7abd52609b");
+    auto& sut = serverPortWithoutOfferOnCreate;
+
+    allocateResponseWithRequestHeaderAndThen(sut, [&](const auto, auto res) {
+        constexpr uint64_t NUMBER_OF_REQUEST_CHUNKS{1U};
+        constexpr uint64_t NUMBER_OF_RESPONSE_CHUNKS{1U};
+        EXPECT_THAT(this->getNumberOfUsedChunks(), Eq(NUMBER_OF_REQUEST_CHUNKS + NUMBER_OF_RESPONSE_CHUNKS));
+        sut.portUser.sendResponse(res);
+        EXPECT_THAT(this->getNumberOfUsedChunks(), Eq(NUMBER_OF_REQUEST_CHUNKS));
+    });
+}
+
+TEST_F(ServerPort_test, SendResponseWithInvalidClientQueueIdReleasesTheChunkToTheMempool)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "45823507-b83b-496b-965e-a48bd3c07b9e");
+    auto& sut = serverPortWithOfferOnCreate;
+
+    // the client is not yet connected to the `clientResponseQueue` which ID is used to send the responses to
+    allocateResponseWithRequestHeaderAndThen(sut, [&](const auto, auto res) {
+        constexpr uint64_t NUMBER_OF_REQUEST_CHUNKS{1U};
+        constexpr uint64_t NUMBER_OF_RESPONSE_CHUNKS{1U};
+        EXPECT_THAT(this->getNumberOfUsedChunks(), Eq(NUMBER_OF_REQUEST_CHUNKS + NUMBER_OF_RESPONSE_CHUNKS));
+        sut.portUser.sendResponse(res);
```

Comment:
After some thinking, I suggest a different API. Since you mentioned you want this to be able to react on potentially lost Responses, e.g. because someone requested a resource, a failed transmission should return a response. The signature of `sendResponse` would become `cxx::expected<ResponseHeader*> sendResponse(ResponseHeader* res);` or `cxx::expected<(ResponseHeader*, ResponseTransmissionFailure)> sendResponse(ResponseHeader* res);`

What do you think?

Reply:
I prefer `cxx::expected<(ResponseHeader*, ResponseTransmissionFailure)> sendResponse(ResponseHeader* res);` Then the error enum would clearly state to other code readers whats going on.

Reply:
I will go this route but with a struct instead of a tuple

Reply:
It seems to be non-trivial to implement this. I added a todo and will implement it in a separate PR since some building blocks need adjustments before the overflow value can be returned